### PR TITLE
fix(executor): `DynamicFilter` cache read strategy

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
`DynamicFilter` cache read strategy

TODO: 
- [X] Add more tests 
- [X] Verify can recover from compute node failure in e2e setting
- [X] minor fixes
- [X] Fix the vnode access issue - update vnodes upon scaling
- [X] Add unit test for `UnorderedRangeCacheIter`

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
fixes: https://github.com/risingwavelabs/risingwave/issues/5726, Partial fix: https://github.com/risingwavelabs/risingwave/issues/6204

### CODE REVIEW

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 1]
```diff
@@ -151,3 +214,150 @@ impl<S: StateStore> RangeCache<S> {
...
+            required_range,
+            false,
+        )
+    } else {
+        // The ranges are non-overlapping. So we delete the old range.
+        (vec![required_range.clone()], required_range, true)
+    }
+}
+
+fn range_contains_lower_upper(
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:321]: add comments

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 2]
```diff
@@ -124,24 +135,66 @@ impl<S: StateStore> RangeCache<S> {
...
+
+                    let vnode_entry = self.cache.entry(vnode).or_insert_with(BTreeMap::new);
+                    while let Some(res) = row_stream.next().await {
+                        let (key_bytes, row) = res?;
+                        // The filter key is always 1st in PK.
+                        let key = deserialize_pk_with_vnode(
+                            &key_bytes.as_ref()[..],
+                            self.state_table.pk_serde(),
+                        )
+                        .unwrap()
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:186]: make this a result via map_err as well.

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 3]
```diff
@@ -124,24 +135,66 @@ impl<S: StateStore> RangeCache<S> {
...
+
+        let to_row_bound = |bound: Bound<ScalarImpl>| -> Bound<Row> {
+            match bound {
+                Unbounded => Unbounded,
+                Included(s) => Included(Row::new(vec![Some(s)])),
+                Excluded(s) => Excluded(Row::new(vec![Some(s)])),
+            }
+        };
+
+        let missing_ranges = missing_ranges
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:161]: Don't need to collect just iter

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 4]
```diff
@@ -151,3 +204,150 @@ impl<S: StateStore> RangeCache<S> {
...
+            let res = iter.next();
+            if res.is_none() {
+                self.current_map = None;
+                self.current_iter = None;
+                self.next()
+            } else {
+                res.map(|r| r.1)
+            }
+        } else {
+            while self.current_map.is_none() && self.next_vnode < u8::MAX {
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:245]: Actually, we don't need to check the first predicate.

**Comment by jon-chuang** [comment]
Unable to reproduce deterministic test unexpected MV output results locally

**Comment by wangrunji0408** [comment]
I'm looking into it and trying to reproduce...

**Comment by jon-chuang** [comment]
Oh wait, maybe I am using HashMap...

**Comment by wangrunji0408** [comment]
HashMap won't be the source of uncertainty...

If your local env is macOS then it's expected that the seed from CI won't work on your local env. Because there are some differences between Linux and Mac, the seed can not be shared.
However, since it failed 2 out of 16 times on CI, you should have a similar failure rate on your local env. It's weird to pass all 256 times. 🤔

**Comment by jon-chuang** [comment]
Let me deliberately introduce a bug and see if it still passes...

**Comment by wangrunji0408** [comment]
I can reproduce this error on my M1 Macbook with:
```
MADSIM_TEST_SEED=15 ./risedev sslt --release -- -j 16 './e2e_test/streaming/**/*.slt'
```

**Comment by jon-chuang** [comment]
Found the bug! But it should be easier to do so than this XD

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 5]
```diff
@@ -151,3 +221,171 @@ impl<S: StateStore> RangeCache<S> {
...
+            let res = iter.next();
+            if res.is_none() {
+                if self.next_vnode == u8::MAX {
+                    // The iterator cannot be refilled further.
+                    self.completed = true;
+                    None
+                } else {
+                    // Try to refill the iterator.
+                    self.next_vnode += 1;
+                    self.current_map = None;
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:282]: Unnecessary step.

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 6]
```diff
@@ -151,3 +221,171 @@ impl<S: StateStore> RangeCache<S> {
...
+    completed: bool,
+    range: ScalarRange,
+}
+
+impl<'a> UnorderedRangeCacheIter<'a> {
+    fn new(
+        cache: &'a HashMap<u8, BTreeMap<ScalarImpl, HashSet<CompactedRow>>>,
+        range: ScalarRange,
+    ) -> Self {
+        Self {
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:239]: We can call refill_iterator on new.

**Comment by jon-chuang** [comment]
Hello, should be ready for final review.

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 7]
```diff
@@ -151,3 +227,370 @@ impl<S: StateStore> RangeCache<S> {
         Ok(())
     }
 }
+
+pub struct UnorderedRangeCacheIter<'a> {
+    cache: &'a HashMap<u8, BTreeMap<ScalarImpl, HashSet<CompactedRow>>>,
```
[the remark this answers is not in the document]
Comment (fuyufjh) [reply src/stream/src/executor/managed_state/dynamic_filter.rs:232]: How about a map from `(sort_key, PK)` to `CompactedRow`?

**Review by fuyufjh — APPROVED** [review]
LGTM for the new cache policy. See my comments.

File: src/stream/src/executor/managed_state/dynamic_filter.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jon-chuang) [reply src/stream/src/executor/managed_state/dynamic_filter.rs:232]: Indeed, that's acceptable.

File: src/stream/src/executor/managed_state/dynamic_filter.rs [hunk 8]
```diff
@@ -151,3 +227,370 @@ impl<S: StateStore> RangeCache<S> {
...
+        for i in 0..=u8::MAX {
+            assert_eq!(
+                Some(&vec![CompactedRow { row: vec![i] }].into_iter().collect()),
+                iter.next()
+            );
+        }
+        assert!(iter.next().is_none());
+    }
+
+    #[test]
```
Comment (jon-chuang) [inline src/stream/src/executor/managed_state/dynamic_filter.rs:566]: Oops, duplicate test. Pls remove.

File: src/stream/src/executor/managed_state/dynamic_filter.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jon-chuang) [reply src/stream/src/executor/managed_state/dynamic_filter.rs:232]: As a matter of fact, we do not even need to encode sort key. We can just use pk. PK is guaranteed to start with the sort key as its the first ordered column.

**Pull request merged**